### PR TITLE
added tests in CategoryTest

### PR DESCRIPTION
**Pull request by mpieron**
I have problems with POST request and checkIfCanFindEntity test

### CODE REVIEW

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 1]
```diff
@@ -62,6 +79,95 @@ public void checkCategoryInAllCategoriesList() {
...
+
+    /**
+     *  Test with posting new category
+     */
+    // Intellij returns status code = 500
+    // "Could not commit JPA transaction; nested exception is javax.persistence.RollbackException: Error while committing the transaction"
+    // When using queryParam instead of CategoryDTO objects, returns code = 400 - same in postman
+
+    @Test
+    public void tryPostNewCategory(){
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:116]: you need to set all the fields in CategoryDTO (except for id, which is set automatically) and it will work

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 2]
```diff
@@ -62,6 +79,95 @@ public void checkCategoryInAllCategoriesList() {
...
+        assertThat(response.getStatusCode()).isEqualTo(200);
+    }
+
+    /**
+     *
+     * Test deleting category
+     */
+    // there should be created new category, posted and then deleted, but post is not working (previous test, code = 400)
+    @Test
+    public void checkIfCanDeleteCategory(){
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:132]: create a new Category as stated in above comment and then delete it and then make sure that you can't get it

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 3]
```diff
@@ -75,8 +181,60 @@ private CategoryDTO getExistingCategoryById(int id) {
...
+                .response();
+    }
+
+    /**
+     * Method accepts Category amd returns Response
+     *
+     * @param newCategory category to create
+     * @return Response object
+     */
+    private Response getResponseFromPostNewCategory(CategoryDTO newCategory){
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:213]: once creating Category will work for you, delete all not needed utility methods

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 4]
```diff
@@ -62,6 +79,95 @@ public void checkCategoryInAllCategoriesList() {
                 hasItem(equalTo(expectedCategory)));
     }
 
+
+
+    /**
+     *  Test with entity search
+     */
+    // it's weird, when extract.as(CategoryDTO[].class) return empty array, at http://localhost:5054/swagger-ui.html it works
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:87]: check how to return ArrayList of given type from rest assured response and see if it will be empty too

File: Vikta-API-testing-sample/src/main/java/com/test/api/execution/BaseTest.java [hunk 5]
```diff
@@ -29,8 +29,13 @@ protected RequestSpecification defaultRequestSpec() {
                 .setAccept(ContentType.JSON).build();
     }
 
-    protected ResponseSpecification defaultResponseSpec() {
+    protected ResponseSpecification defaultResponseJsonSpec() {
         return new ResponseSpecBuilder()
                 .expectContentType(ContentType.JSON).build();
     }
+
+    protected ResponseSpecification defaultResponseTxtSpec() {
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/main/java/com/test/api/execution/BaseTest.java:37]: once creating will work, please delete this method as it should not be needed

**Review by kdyla-griddynamics — CHANGES_REQUESTED** [review]

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 6]
```diff
@@ -20,6 +22,7 @@
      */
     @Test
     public void verifyGetCategory() {
+
         final int categoryId = 5;
         final String expectedTitle = "Street";
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:28]: I know it's not the message You set, but please rework all the assertions messages in following manner:
"Category title should be %s, but is %s"
Using this approach gives a clearer understanding of what is expected and what is the actual outcome when assertion fails. Please apply to all assertions

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 7]
```diff
@@ -28,17 +31,29 @@ public void verifyGetCategory() {
                 equalTo(expectedTitle));
     }
 
+    /**
+     * Test with wrong id given
+     */
+    @Test
+    public void verifyGetCategoryWithUnexpectedId() {
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:38]: create a parameterized tests for negative cases. for each available action create separate test method that would accept several negative cases to assert. Use documentation to check boundary values and prepare cases based on that

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 8]
```diff
@@ -28,17 +31,29 @@ public void verifyGetCategory() {
...
+
+        assertThat("Category id is correct, excepted wrong", response.getStatusCode(), equalTo(404));
+        assertThat(response.getBody().asString(), containsString("No such Category entity"));
+    }
+
     /**
      * Test with wrong description given
      */
     @Test
     public void verifyGetCategoryWithUnexpectedDescription() {
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:51]: this could one of the cases for comment above

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 9]
```diff
@@ -62,6 +77,170 @@ public void checkCategoryInAllCategoriesList() {
...
     }
 
+    /**
+     *  Test with entity search
+     */
+    // city is empty list
+    @Test
+    public void checkIfCanFindEntity(){
+        final String termQueryParam = "term";
+        final String term = "Wolf%20N8%20%2F%20Dynamic%20Marketing%20Officer";
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:87]: please list the available categories first and set such search term that would return more than one category. I copied the example from 'Try it out' in the docs and it's working

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 10]
```diff
@@ -62,6 +77,170 @@ public void checkCategoryInAllCategoriesList() {
...
+
+        List<CategoryDTO> city;
+        city = given().spec(defaultRequestSpec())
+                .queryParam(termQueryParam, term)
+                .when()
+                .log().all()
+                .get(getTestEnvironment().getCategoryPath() + "/search")
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:94]: store search uri path in a variable

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 11]
```diff
@@ -79,4 +258,36 @@ private CategoryDTO getExistingCategoryById(int id) {
...
+                .extract()
+                .response();
+    }
+
+    /**
+     * Method used to clean after test that creates new Category
+     *
+     * @param id category identifier
+     */
+    private void cleanAfterTest(int id){
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:285]: try to make this method accept 'long' so You wouldn't have to downcast to 'int' every time You use it

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 12]
```diff
@@ -79,4 +258,36 @@ private CategoryDTO getExistingCategoryById(int id) {
...
                 .as(CategoryDTO.class);
     }
+
+    /**
+     * Method returns Response and expects that category exist
+     *
+     * @param id category identifier
+     * @return Response object
+     */
+    private Response getResponseFromCategoryById(int id) {
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:268]: try to make this method accept 'long' so You wouldn't have to downcast to 'int' every time You use it

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 13]
```diff
@@ -62,6 +77,170 @@ public void checkCategoryInAllCategoriesList() {
...
+                        .log().all()
+                        .when()
+                        .post(getTestEnvironment().getCategoryPath())
+                        .then()
+                        .log().all()
+                        .spec(defaultResponseSpec())
+                        .extract().as(CategoryDTO.class);
+
+        Response responseGet = getResponseFromCategoryById((int)createdCategory.getId());
+        cleanAfterTest((int)createdCategory.getId());
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:129]: try to move the cleaning part to some @After method

**Review by kdyla-griddynamics — CHANGES_REQUESTED** [review]
We got more and more working methods and that's great. Now we need to make sure that the API is tested really thoroughly. Check the ranges of the values accepted by each method and design test cases that would check positive and negative cases for each action. for negative cases it's best to use parameterized tests. for positive - it's up to You :)

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 14]
```diff
@@ -2,50 +2,63 @@
...
+
+
     @Test
-    public void verifyGetCategory() {
+    public void canGetCategoryByCorrectId() {
         final int categoryId = 5;
         final String expectedTitle = "Street";
 
-        assertThat("Category title is unexpected",
+        assertThat("Category title should be %s, but is %s",
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:36]: this expression should use String.format, otherwise when test fails, the message is: 
java.lang.AssertionError: Category title should be %s, but is %s

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 15]
```diff
@@ -2,50 +2,63 @@
...
+        final  int categoryId = 0;
+
+        Response response = getResponseFromCategoryById(categoryId);
+
+        assertThat(String.format("Category id: %s is correct, excepted wrong" ,categoryId), response.getStatusCode(), equalTo(404));
+        assertThat("There shouldn't be such a category" + categoryId, response.getBody().asString(), containsString("No such Category entity"));
+    }
+
+    @Test
+    public void canNotGetCategoryWithUnexpectedDescription() {
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:52]: this test does not make a lot of sense. I know it was there at the start, but we should get rid of it. Basically, You're getting some category by existing id and asserting that its description is not equal to some made up String

**Review by kdyla-griddynamics — CHANGES_REQUESTED** [review]
we're getting there :)
first thing: there's a couple of important cases missing - negative create category, for example 'you cannot create category without setting title' etc., negative update, and boundary values that we discussed yesterday.
second thing: you use two types of assertions from different libraries. I would stick to assertj, because they are more readable

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 16]
```diff
@@ -57,19 +78,234 @@ public void checkCategoryInAllCategoriesList() {
...
                 .as(CategoryDTO[].class);
 
-        assertThat("Category wasn't found in all categories list",
+        assertThat(String.format("Category %s wasn't found in all categories list", categoryId),
                 Arrays.asList(categories),
                 hasItem(equalTo(expectedCategory)));
     }
 
+    @Test
+    public void canListAllCategories(){
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:87]: in this test, You just get the same list twice and assert that they're equal. Get it once and assert that it's not empty, and that all elements have required fields

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 17]
```diff
@@ -57,19 +78,234 @@ public void checkCategoryInAllCategoriesList() {
...
+                .when()
+                .get(pathList)
+                .then()
+                .spec(defaultResponseSpec())
+                .extract()
+                .body().jsonPath().getList(".", CategoryDTO.class);
+
+        StringBuilder termBuilder = new StringBuilder();
+        for(int i=0; i<size; i++){
+            termBuilder.append(listOfCategories.get(random.nextInt(listOfCategories.size())).getTitle());
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:127]: if You'll pick the same category twice - which is possible with random(), the test will fail

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 18]
```diff
@@ -57,19 +78,234 @@ public void checkCategoryInAllCategoriesList() {
...
+                .extract()
+                .body().jsonPath().getList(".", CategoryDTO.class);
+        System.out.println(term);
+        for(CategoryDTO cat : city)
+            System.out.println(cat.getTitle());
+
+        assertThat(String.format("Couldn't find all entities. Found only %d", city.size()), city.size() == size);
+    }
+
+    @ParameterizedTest(name = "{index} => title={0}, description={1}, pathToCatImage={2}, imageItemIds={3}")
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:151]: you don't need the expression from the brackets

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 19]
```diff
@@ -57,19 +78,234 @@ public void checkCategoryInAllCategoriesList() {
...
+        Response responseGet = getResponseFromCategoryById(createdCategory.getId());
+        idCategoryToClean.add(createdCategory.getId());
+
+        assertThat(createdCategory).usingRecursiveComparison()
+                .ignoringFields(idQueryParam, "imageItemIds")
+                .isEqualTo(newCategory);
+        assertThat(responseGet.getStatusCode()).isEqualTo(200);
+    }
+
+    @ParameterizedTest(name = "{index} => title={0}, description={1}, pathToCatImage={2}, imageItemIds={3}")
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:182]: same as above

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 20]
```diff
@@ -79,4 +315,46 @@ private CategoryDTO getExistingCategoryById(int id) {
...
+        return given().spec(defaultRequestSpec())
+                .when().queryParam(idQueryParam, id)
+                .get(getTestEnvironment().getCategoryPath())
+                .then()
+                .spec(defaultResponseSpec())
+                .extract()
+                .response();
+    }
+
+    private static Stream<Arguments> goodParametersForPost() {
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:336]: add test description as fifth element, and log it at the beginning of the test

**Review by kdyla-griddynamics — CHANGES_REQUESTED** [review]

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 21]
```diff
@@ -2,53 +2,65 @@
...
-    public void verifyGetCategory() {
+    public void canGetCategoryByCorrectId() {
         final int categoryId = 5;
         final String expectedTitle = "Street";
 
-        assertThat("Category title is unexpected",
-                getExistingCategoryById(categoryId).getTitle(),
-                equalTo(expectedTitle));
+        assertThat(getExistingCategoryById(categoryId).getTitle())
+                .as(() -> String.format("Category title should be %s, but is %s",getExistingCategoryById(categoryId).getTitle(), expectedTitle))
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:47]: You don't have to pass the message as lambda expression, it can be String. Same for all messages below

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 22]
```diff
@@ -57,9 +69,228 @@ public void checkCategoryInAllCategoriesList() {
                 .extract()
                 .as(CategoryDTO[].class);
 
-        assertThat("Category wasn't found in all categories list",
-                Arrays.asList(categories),
-                hasItem(equalTo(expectedCategory)));
+        assertThat(categories).as(() -> "Category list shouldn't be empty").isNotEmpty();
+
+        for(CategoryDTO category : categories){
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:74]: Assertions in a loop should be avoided. Find an assertion that checks the content of the list or use SoftAssertions.

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 23]
```diff
@@ -57,9 +69,228 @@ public void checkCategoryInAllCategoriesList() {
...
+
+        List <CategoryDTO> listOfCategories = given().spec(defaultRequestSpec())
+                .when()
+                .get(pathList)
+                .then()
+                .spec(defaultResponseSpec())
+                .extract()
+                .body().jsonPath().getList(".", CategoryDTO.class);
+
+        assert listOfCategories.size() > 0;
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:98]: please use assertion from chosen library

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 24]
```diff
@@ -57,9 +69,228 @@ public void checkCategoryInAllCategoriesList() {
...
+                .queryParam(termQueryParam, term)
+                .when()
+                .log().all()
+                .get(searchPath)
+                .then()
+                .log().all()
+                .spec(defaultResponseSpec())
+                .extract()
+                .body().jsonPath().getList(".", CategoryDTO.class);
+        System.out.println(size);
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:119]: remove printouts after debugging

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 25]
```diff
@@ -57,9 +69,228 @@ public void checkCategoryInAllCategoriesList() {
...
+
+        int size = random.nextInt(listOfCategories.size()) + 1;
+
+        String term =  listOfCategories.stream()
+                .map(CategoryDTO::getTitle)
+                .limit(size)
+                .map(x -> x.replaceAll(" ", "~"))
+                .collect(Collectors.joining("|"));
+
+        List<CategoryDTO> city;
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:108]: please give some more descriptive name, e.g resultList

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 26]
```diff
@@ -57,9 +69,228 @@ public void checkCategoryInAllCategoriesList() {
...
+            System.out.println(cat.getTitle());
+
+        assertThat(listOfCategories)
+                .as("Category list is empty, can't check if can find entity")
+                .isNotEmpty();
+        assertThat(city)
+                .as("Category search result list shouldn't be empty")
+                .isNotEmpty();
+        assertThat(city.size())
+                .as("Couldn't find all entities")
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:135]: description should be more like: There should be %s results, but got only %s

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 27]
```diff
@@ -57,9 +69,228 @@ public void checkCategoryInAllCategoriesList() {
...
+        assertThat(response.getStatusCode()).isEqualTo(404);
+    }
+
+    @Test
+    @Tag("needs-cleanup")
+    public void canUpdateExistingCategory(){
+        Set<Long> set = new HashSet<>();
+        set.add(1000L);
+        String newTitle = "Updated Category";
+        int id;
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:256]: you can declare this variable closer to the place where it's used

File: Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java [hunk 28]
```diff
@@ -57,9 +69,228 @@ public void checkCategoryInAllCategoriesList() {
...
+        assertThat(createdCategory).usingRecursiveComparison()
+                .ignoringFields(idQueryParam, "imageItemIds")
+                .isEqualTo(newCategory);
+        assertThat(responseGet.getStatusCode()).isEqualTo(200);
+    }
+
+    @ParameterizedTest(name = "{index} => title={0}, description={1}, pathToCatImage={2}, imageItemIds={3}")
+    @MethodSource("wrongParametersForPost")
+    public void canNotPostNewCategoryWithWrongParameters(String title, String description, String pathToCatImage, long imageItemIds, String message){
+        logger.error(message);
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/category/CategoryTest.java:174]: log this at info level, because it's just a description

**Review by kdyla-griddynamics — CHANGES_REQUESTED** [review]

**Review by kdyla-griddynamics — APPROVED** [review]
